packages/core/src/authoring/model/model_signal.ts

```
@@ -0,0 +1,117 @@
+   * name as the input, but suffixed with `Change`. By default, the class field name is used.
+   */
+  alias?: string;
+}
+
+/**
+ * `ModelSignal` represents a special `Signal` for a directive/component model field.
+ *
+ * A model signal is a writeable signal that can be exposed as an output.
+ * Whenever its value is updated, it emits to the output.
+ *
+ * @developerPreview
+ */
+export interface ModelSignal<T> extends WritableSignal<T> {
+  [SIGNAL]: ModelSignalNode<T>;
+  [ɵINPUT_SIGNAL_BRAND_READ_TYPE]: T;
+  [ɵINPUT_SIGNAL_BRAND_WRITE_TYPE]: T;
+
+  /**
+   * Subscribes to changes in the model's value.
+   * @internal
+   */
+  subscribe(callback: (value: T) => void): {unsubscribe: () => void};
+}
+
+/**
+ * Creates a model signal.
+ *
+ * @param initialValue The initial value.
+ *   Can be set to {@link REQUIRED_UNSET_VALUE} for required model signals.
+ * @param options Additional options for the model.
+ */
+export function createModelSignal<T>(initialValue: T): ModelSignal<T> {
+  const node: ModelSignalNode<T> = Object.create(MODEL_SIGNAL_NODE);
+
+  node.value = initialValue;
+
+  function getter(): T {
+    producerAccessed(node);
+
+    if (node.value === REQUIRED_UNSET_VALUE) {
+      throw new RuntimeError(
+          RuntimeErrorCode.REQUIRED_MODEL_NO_VALUE,
+          ngDevMode && 'Model is required but no value is available yet.');
+    }
+
+    return node.value;
+  }
+
+  function notifySubscribers(value: T): void {
+    const subscriptions = node.subscriptions;
+
+    for (let i = 0; i < subscriptions.length; i++) {
+      subscriptions[i](value);
+    }
+  }
+
+  (getter as any)[SIGNAL] = node;
+  (getter as any).asReadonly = (() => getter()) as Signal<T>;
+
+  getter.set = (newValue: T) => {
+    if (!node.equal(node.value, newValue)) {
+      signalSetFn(node, newValue);
+      notifySubscribers(newValue);
+    }
+  };
+
+  getter.update = (updateFn: (value: T|typeof REQUIRED_UNSET_VALUE) => T) => {
```

Does this `typeof` need to be here? We'll never call the user's function with it afaik. Calling `.update` when the signal isn't set in the first place should 💥

---

It is, otherwise TS produces the following:

> Argument of type 'T | unique symbol' is not assignable to parameter of type 'T'.

---

It's clearly wrong as a type though, which suggests that something is possible that shouldn't be. We should never be calling the user's `update` function with `REQUIRED_UNSET_VALUE`.

If the user calls `.update()` on an unset `model()`, it should throw the same error as if they tried to read it. We should have a test for this.

---

Fixed.